In [51]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression

# Importer le dataframe coeur

In [52]:
data = pd.read_excel('Copie de Coeur.xlsx')
df = data.copy()

# Preprocessing


In [53]:
#normalisé les variables quantitatives

def normalize_quant_var(df):
    #quant = df.select_dtypes('int')
    for i in df.select_dtypes('int'):
        df[i]= df[i] / df[i].max()
    return df
        

normalize_quant_var(data)

data


AGE   SEXE  TDT   PAR  CHOLESTEROL  GAJ     ECG     FCMAX ANGINE  \
0    0.519481  homme   AA  0.70     0.479270  0.0  Normal  0.851485    Non   
1    0.636364  femme  DNA  0.80     0.298507  0.0  Normal  0.772277    Non   
2    0.480519  homme   AA  0.65     0.469320  0.0      ST  0.485149    Non   
3    0.623377  femme  ASY  0.69     0.354892  0.0  Normal  0.534653    Oui   
4    0.701299  homme  DNA  0.75     0.323383  0.0  Normal  0.603960    Non   
..        ...    ...  ...   ...          ...  ...     ...       ...    ...   
913  0.584416  homme   AT  0.55     0.437811  0.0  Normal  0.653465    Non   
914  0.883117  homme  ASY  0.72     0.320066  1.0  Normal  0.698020    Non   
915  0.740260  homme  ASY  0.65     0.217247  0.0  Normal  0.569307    Oui   
916  0.740260  femme   AA  0.65     0.391376  0.0     LVH  0.861386    Non   
917  0.493506  homme  DNA  0.69     0.290216  0.0  Normal  0.856436    Non   

     DEPRESSION       PENTE  CŒUR  
0            0.0  Ascendant   0.0  
1            1.0       Plat   1.0  
2            0.0  Ascendant   0.0  
3            1.5       Plat   1.0  
4            0.0  Ascendant   0.0  
..           ...        ...   ...  
913          1.2       Plat   1.0  
914          3.4       Plat   1.0  
915          1.2       Plat   1.0  
916          0.0       Plat   1.0  
917          0.0  Ascendant   0.0  

[918 rows x 12 columns]

In [ ]:
#FILTRE

#notre dataframe ne contient pas de doublons 
data.duplicated().value_counts()
#data ne contient pas de données vide
data.isna().value_counts()
#il y a pas de constantes
data.nunique()

In [54]:
#recoder les variables qualitatives
def recode_quality_var(serie_name):
     
    return serie_name.astype('category').cat.codes

def data_recode_qual(df):
    for col in df.select_dtypes('object').columns:
        df[col]=recode_quality_var(df[col])
    return df
    
data = data_recode_qual(data)
data

AGE  SEXE  TDT   PAR  CHOLESTEROL  GAJ  ECG     FCMAX  ANGINE  \
0    0.519481     1    0  0.70     0.479270  0.0    1  0.851485       0   
1    0.636364     0    3  0.80     0.298507  0.0    1  0.772277       0   
2    0.480519     1    0  0.65     0.469320  0.0    2  0.485149       0   
3    0.623377     0    1  0.69     0.354892  0.0    1  0.534653       1   
4    0.701299     1    3  0.75     0.323383  0.0    1  0.603960       0   
..        ...   ...  ...   ...          ...  ...  ...       ...     ...   
913  0.584416     1    2  0.55     0.437811  0.0    1  0.653465       0   
914  0.883117     1    1  0.72     0.320066  1.0    1  0.698020       0   
915  0.740260     1    1  0.65     0.217247  0.0    1  0.569307       1   
916  0.740260     0    0  0.65     0.391376  0.0    0  0.861386       0   
917  0.493506     1    3  0.69     0.290216  0.0    1  0.856436       0   

     DEPRESSION   PENTE  CŒUR  
0            0.0      0   0.0  
1            1.0      2   1.0  
2            0.0      0   0.0  
3            1.5      2   1.0  
4            0.0      0   0.0  
..           ...    ...   ...  
913          1.2      2   1.0  
914          3.4      2   1.0  
915          1.2      2   1.0  
916          0.0      2   1.0  
917          0.0      0   0.0  

[918 rows x 12 columns]

In [55]:
#deux objet X[age:pente] Y[coeur]

X = data.drop('CŒUR', axis= 1)
print(X)
# on recupere pour Y  df['CŒUR']  qui es les valeurs de coeur non mormalisées
Y = df['CŒUR']
Y


          AGE  SEXE  TDT   PAR  CHOLESTEROL  GAJ  ECG     FCMAX  ANGINE  \
0    0.519481     1    0  0.70     0.479270  0.0    1  0.851485       0   
1    0.636364     0    3  0.80     0.298507  0.0    1  0.772277       0   
2    0.480519     1    0  0.65     0.469320  0.0    2  0.485149       0   
3    0.623377     0    1  0.69     0.354892  0.0    1  0.534653       1   
4    0.701299     1    3  0.75     0.323383  0.0    1  0.603960       0   
..        ...   ...  ...   ...          ...  ...  ...       ...     ...   
913  0.584416     1    2  0.55     0.437811  0.0    1  0.653465       0   
914  0.883117     1    1  0.72     0.320066  1.0    1  0.698020       0   
915  0.740260     1    1  0.65     0.217247  0.0    1  0.569307       1   
916  0.740260     0    0  0.65     0.391376  0.0    0  0.861386       0   
917  0.493506     1    3  0.69     0.290216  0.0    1  0.856436       0   

     DEPRESSION   PENTE  
0            0.0      0  
1            1.0      2  
2            0.0     

0      0
1      1
2      0
3      1
4      0
      ..
913    1
914    1
915    1
916    1
917    0
Name: CŒUR, Length: 918, dtype: int64

# Apprentissage


In [64]:
#Decouper en apprentissage et en texte
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size=0.3, random_state = 0)

#initialisation du model de regression
model = LogisticRegression()

#Entrainement du model
model.fit(x_train,y_train)

#Verifions le taux d'exatitude de notre model 
model.score(x_test,y_test)

0.8297101449275363

# Interpretation des perfommance du model

In [57]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

In [58]:
# La matrice de confusion.

matrice_conf = confusion_matrix(y_test, model.predict(x_test))

#La matrice de confusion est une confrontation entre les valeurs réelles de la variable cible et les
#valeurs prédites par le modèle.

matrice_conf

array([[ 90,  23],
       [ 24, 139]])

# Remarques

### le nombre de vrais négatifs est M(0,0) = matrice_conf[0,0] |  dans notre cas VN = 90
### les faux négatifs sont M(1,0) = matrice_conf[1,0]          |  dans notre cas FN = 24
### les vrais positifs sont M(1,1) = matrice_conf[1,1]         |  dans notre cas VP = 139
### les faux positifs sont (0,1) = matrice_conf[0,1]           |  dans notre cas FP = 23 


In [60]:
#la sensibilité


recall_score(y_test,model.predict(x_test))

#Lorsque individu est positif on a 85.27 % de chance que le modèle arrive à bien le prédire (predire qu'il est positif)

#formule mathematique : sensibilité = VP / (VP + FP)

0.852760736196319

In [61]:
#precision 

precision_score(y_test,model.predict(x_test))

#Lorsque le modèle prédit individu est positif on a 85.80 % de chance que cela soit vrai.

#formule mathematique : prediction = VP / (VP + FN)

0.8580246913580247

In [62]:
#Calculons le taux de succes de notre model d'apres la fomule suivante : 

accuracy = (matrice_conf[0,0]  + matrice_conf[1,1]) / (matrice_conf[1,1] + matrice_conf[0,0] + matrice_conf[1,0] + matrice_conf[0,1] )
#Multiplions le resultat par 100 pour l'avoire en pourcentage 
accuracy*100

82.97101449275362

# Conclusion 


### D'apres nôtre accuracy (le taux de succes de notre model ) qui est d'environ 0.8297 soit 82.97% 
### Nous pouvons dit que nous avons un bon model , Cependant des questions restent posées .
### Avons nous eviter tout les biais liés a ce model ?
### Le model est t-il le meilleure possible ?
